# Домашнее задание к лекции "Статистика. Практика"

## Задание 1

Вернемся к [набору данных о видеоиграх](https://github.com/obulygin/pyda_homeworks/blob/master/stat_case_study/video_games_sales.csv).

Ответьте на следующие вопросы:

1) Как критики относятся к спортивным играм?  
2) Критикам нравятся больше игры на PC или на PS4?  
3) Критикам больше нравятся стрелялки или стратегии?  

Для каждого вопроса:
- сформулируйте нулевую и альтернативную гипотезы;
- выберите пороговый уровень статистической значимости;
- опишите полученные результаты статистического теста.

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import scipy.stats as stats

In [2]:
df = pd.read_csv('video_games_sales.csv')

In [3]:
df.head()

Rank                           Name                      basename  \
0     1                     Wii Sports                    wii-sports   
1     2              Super Mario Bros.              super-mario-bros   
2     3                 Mario Kart Wii                mario-kart-wii   
3     4  PlayerUnknown's Battlegrounds  playerunknowns-battlegrounds   
4     5              Wii Sports Resort             wii-sports-resort   

      Genre ESRB_Rating Platform         Publisher         Developer  \
0    Sports           E      Wii          Nintendo      Nintendo EAD   
1  Platform         NaN      NES          Nintendo      Nintendo EAD   
2    Racing           E      Wii          Nintendo      Nintendo EAD   
3   Shooter         NaN       PC  PUBG Corporation  PUBG Corporation   
4    Sports           E      Wii          Nintendo      Nintendo EAD   

   VGChartz_Score  Critic_Score  ...  NA_Sales  PAL_Sales  JP_Sales  \
0             NaN           7.7  ...       NaN        NaN       NaN   
1             NaN          10.0  ...       NaN        NaN       NaN   
2             NaN           8.2  ...       NaN        NaN       NaN   
3             NaN           NaN  ...       NaN        NaN       NaN   
4             NaN           8.0  ...       NaN        NaN       NaN   

   Other_Sales    Year  Last_Update  \
0          NaN  2006.0          NaN   
1          NaN  1985.0          NaN   
2          NaN  2008.0  11th Apr 18   
3          NaN  2017.0  13th Nov 18   
4          NaN  2009.0          NaN   

                                                 url  status Vgchartzscore  \
0  http://www.vgchartz.com/game/2667/wii-sports/?...       1           NaN   
1  http://www.vgchartz.com/game/6455/super-mario-...       1           NaN   
2  http://www.vgchartz.com/game/6968/mario-kart-w...       1           8.7   
3  http://www.vgchartz.com/game/215988/playerunkn...       1           NaN   
4  http://www.vgchartz.com/game/24656/wii-sports-...       1           8.8   

                                         img_url  
0  /games/boxart/full_2258645AmericaFrontccc.jpg  
1                   /games/boxart/8972270ccc.jpg  
2  /games/boxart/full_8932480AmericaFrontccc.jpg  
3  /games/boxart/full_8052843AmericaFrontccc.jpg  
4  /games/boxart/full_7295041AmericaFrontccc.jpg  

[5 rows x 23 columns]

#### 1) Как критики относятся к спортивным играм?
Будем считать, что критики относятся положительно к играм, если среднее значение выставленного ими балла больше и равно 7,5.

Тогда, 

* H0: средний выставленный критиками балл для спортивной игры >= 7,5
* H1: средний выставленный критиками балл для спортивной игры < 7,5

Сравниваем значение по выборке с константой, поэтому нужен одновыборочный критерий. Тест нужен односторонний, т.к. нас интересует только неравенство в меньшую сторону.

In [4]:
alpha = 0.05
result = stats.ttest_1samp(df[df.Genre == 'Sports'].Critic_Score, 7.5, nan_policy='omit')
print(result)
if (result.statistic < 0) & (result.pvalue/2 < alpha): 
    print('Отвергаем нулевую гипотезу, среднее значение балла критиков меньше 7.5')
else:
    print('Не отвергаем нулевую гипотезу, критикам нравятся спортивные игры')

Ttest_1sampResult(statistic=-3.719660933773832, pvalue=0.00021577789041611826)
Отвергаем нулевую гипотезу, среднее значение балла критиков меньше 7.5


#### Итог:
Критики негативно относятся к спортивным играм, т.к. среднее значение балла критиков меньше 7.5.

Пороговый уровень статистической значимости равен pvalue=0.000107

#### 2) Критикам нравятся больше игры на PC или на PS4?

* H0: критикам нравятся игры на PC также, как и на PS4
* H1: критикам нравятся игры на PC больше/меньше, чем на PS4

Сравниваем две выборки, поэтому нужен двухвыборочный двухсторонний критерий.

In [5]:
alpha = 0.05
result = stats.ttest_ind(df[df.Platform == 'PC'].Critic_Score, df[df.Platform == 'PS4'].Critic_Score, equal_var=False, nan_policy='omit')
print(result)
if result.pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, критикам нравятся игры на PC и на PS4 по-разному')
else:
    print('Не отвергаем нулевую гипотезу, критикам нравятся игры на PC и на PS4 одинаково')

Ttest_indResult(statistic=-2.7394476056351627, pvalue=0.0069318082502542845)
Отвергаем нулевую гипотезу, критикам нравятся игры на PC и на PS4 по-разному


#### Итог:
Критикам нравятся игры на PS4 больше чем, на PC, так как статистический коэффициент меньше нуля.

Пороговый уровень статистической значимости равен pvalue=0.0069.

#### 3) Критикам больше нравятся стрелялки или стратегии?
* H0: стрелялки и стратегии нравятся критикам одинаково.
* H1: стрелялки критикам нравятся больше/меньше, чем стратегии.

Сравниваем две выборки, поэтому нужен двухвыборочный двухсторонний критерий.

In [6]:
alpha = 0.05
result = stats.ttest_ind(df[df.Genre == 'Shooter'].Critic_Score, df[df.Genre == 'Strategy'].Critic_Score, equal_var=False, nan_policy='omit')
print(result)
if result.pvalue < alpha: 
    print('Отвергаем нулевую гипотезу, стрелялки и стратегии критики оценивает по-разному')
else:
    print('Нет оснований отвергать нулевую гипотезу, стрелялки и стратегии нравятся критикам одинаково')

Ttest_indResult(statistic=-1.6073949711166526, pvalue=0.10838786414222695)
Нет оснований отвергать нулевую гипотезу, стрелялки и стратегии нравятся критикам одинаково


#### Итог:
Стрелялки и стратегии нравятся критикам одинаково.
Пороговый уровень статистической значимости равен pvalue=0.108

## Задание 2

Реализуйте базовую модель логистической регрессии для классификации текстовых сообщений (используемые данные [здесь](https://github.com/obulygin/pyda_homeworks/blob/master/stat_case_study/spam.csv)) по признаку спама. Для этого:

1) Привидите весь текст к нижнему регистру;  
2) Удалите мусорные символы;  
3) Удалите стоп-слова;  
4) Привидите все слова к нормальной форме;  
5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:  

```
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=cname)
```

Можете поэкспериментировать с параметрами [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html);  
6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите `random_state=42`. Используйте [train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html);  
7) Постройте модель [логистической регрессии](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html), укажите `random_state=42`, оцените ее точность на тестовых данных;  
8) Опишите результаты при помощи [confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html?highlight=confusion_matrix#sklearn.metrics.confusion_matrix);  
9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [7]:
import re

In [8]:
spam = pd.read_csv('spam.csv')
spam.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [9]:
spam_mod = spam.copy(deep=True)

#### 1) Привидите весь текст к нижнему регистру

In [10]:
spam_mod.Message = spam_mod.Message.str.lower()
spam_mod

Category                                            Message
0         ham  go until jurong point, crazy.. available only ...
1         ham                      ok lar... joking wif u oni...
2        spam  free entry in 2 a wkly comp to win fa cup fina...
3         ham  u dun say so early hor... u c already then say...
4         ham  nah i don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  this is the 2nd time we have tried 2 contact u...
5568      ham               will ü b going to esplanade fr home?
5569      ham  pity, * was in mood for that. so...any other s...
5570      ham  the guy did some bitching but i acted like i'd...
5571      ham                         rofl. its true to its name

[5572 rows x 2 columns]

#### 2) Удалите мусорные символы

In [11]:
spam_mod.Message = spam_mod['Message'].apply(lambda row: re.sub('[\W_]+',' ', row))
spam_mod

Category                                            Message
0         ham  go until jurong point crazy available only in ...
1         ham                           ok lar joking wif u oni 
2        spam  free entry in 2 a wkly comp to win fa cup fina...
3         ham       u dun say so early hor u c already then say 
4         ham  nah i don t think he goes to usf he lives arou...
...       ...                                                ...
5567     spam  this is the 2nd time we have tried 2 contact u...
5568      ham               will ü b going to esplanade fr home 
5569      ham  pity was in mood for that so any other suggest...
5570      ham  the guy did some bitching but i acted like i d...
5571      ham                          rofl its true to its name

[5572 rows x 2 columns]

#### 3) Удалите стоп-слова

In [12]:
from nltk.corpus import stopwords

In [13]:
stopwords_set = set(stopwords.words('english'))

In [14]:
def clean_stop_words(row):
    text = row.split()
    text_without_sw = [word for word in text if word not in stopwords_set]
    return ' '.join(text_without_sw)

In [15]:
spam_mod.Message = spam_mod['Message'].apply(clean_stop_words)
spam_mod

Category                                            Message
0         ham  go jurong point crazy available bugis n great ...
1         ham                            ok lar joking wif u oni
2        spam  free entry 2 wkly comp win fa cup final tkts 2...
3         ham                u dun say early hor u c already say
4         ham             nah think goes usf lives around though
...       ...                                                ...
5567     spam  2nd time tried 2 contact u u 750 pound prize 2...
5568      ham                        ü b going esplanade fr home
5569      ham                              pity mood suggestions
5570      ham  guy bitching acted like interested buying some...
5571      ham                                     rofl true name

[5572 rows x 2 columns]

#### 4) Привидите все слова к нормальной форме;

In [16]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()

In [17]:
def lemmatize_text(row):
    text = row.split()
    lemmatized_text = [wordnet_lemmatizer.lemmatize(word) for word in text]
    return ' '.join(lemmatized_text)

In [18]:
spam_mod.Message = spam_mod['Message'].apply(lemmatize_text)
spam_mod

Category                                            Message
0         ham  go jurong point crazy available bugis n great ...
1         ham                            ok lar joking wif u oni
2        spam  free entry 2 wkly comp win fa cup final tkts 2...
3         ham                u dun say early hor u c already say
4         ham                nah think go usf life around though
...       ...                                                ...
5567     spam  2nd time tried 2 contact u u 750 pound prize 2...
5568      ham                        ü b going esplanade fr home
5569      ham                               pity mood suggestion
5570      ham  guy bitching acted like interested buying some...
5571      ham                                     rofl true name

[5572 rows x 2 columns]

#### 5) Преобразуйте все сообщения в вектора TF-IDF. 

In [19]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(spam_mod.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=names)

In [20]:
tfidf_matrix

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
...   ...  ...     ...           ...   ...   ...          ...          ...   
5567  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5568  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5569  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5570  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5571  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

      0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
1            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
2            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
3            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
4            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
...          ...  ...  ...    ...     ...  ...        ...   ...   ...    ...   
5567         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5568         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5569         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5570         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5571         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   

       èn   ú1  〨ud  
0     0.0  0.0  0.0  
1     0.0  0.0  0.0  
2     0.0  0.0  0.0  
3     0.0  0.0  0.0  
4     0.0  0.0  0.0  
...   ...  ...  ...  
5567  0.0  0.0  0.0  
5568  0.0  0.0  0.0  
5569  0.0  0.0  0.0  
5570  0.0  0.0  0.0  
5571  0.0  0.0  0.0  

[5572 rows x 8048 columns]

6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. Используйте train_test_split;

7) Постройте модель логистической регрессии, укажите random_state=42, оцените ее точность на тестовых данных;


In [21]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(tfidf_matrix, spam_mod.Category, test_size=0.30, random_state=42)

In [22]:
from sklearn.linear_model import LogisticRegression

In [23]:
LR = LogisticRegression(random_state=42)

In [24]:
LR.fit(X_train, Y_train)

LogisticRegression(random_state=42)

In [25]:
LR.predict(X_test)

array(['ham', 'ham', 'ham', ..., 'ham', 'spam', 'ham'], dtype=object)

In [27]:
result = pd.DataFrame([Y_test.index, Y_test.to_numpy(), LR.predict(X_test)]).T

In [28]:
result.set_index(0, inplace=True)
result.columns = ['Category','Predicted_Category']
result

Category Predicted_Category
0                               
3245      ham                ham
944       ham                ham
1044      ham                ham
2484      ham                ham
812       ham                ham
...       ...                ...
2505      ham                ham
2525     spam               spam
4975      ham                ham
650      spam               spam
4463      ham                ham

[1672 rows x 2 columns]

In [30]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_test, LR.predict(X_test))

0.958732057416268

8) Опишите результаты при помощи confusion_matrix;

In [31]:
from sklearn.metrics import confusion_matrix

In [32]:
confusion_matrix(Y_test, LR.predict(X_test))

array([[1445,    3],
       [  66,  158]], dtype=int64)

1445+158 верно предсказанных значения, 66+3 неправильно предсказанных

In [33]:
tn, fp, fn, tp = confusion_matrix(Y_test, LR.predict(X_test)).ravel()

In [34]:
accuracy = (tp + tn) / (tn + fp + fn + tp)
print('Точность классификации: %.2f ' % accuracy)
recall = (tp) / (fn + tp)
print('Чувствительность - отношение общего количества правильно классифицированных положительных классов \n к общему количеству положительных классов: %.2f' %recall)
precision = (tp) / (fp + tp)
print('Точность - отношение общего количества правильно классифицированных положительных классов \n к общему количеству предсказанных положительных классов: %.2f ' % precision )

Точность классификации: 0.96 
Чувствительность - отношение общего количества правильно классифицированных положительных классов 
 к общему количеству положительных классов: 0.71
Точность - отношение общего количества правильно классифицированных положительных классов 
 к общему количеству предсказанных положительных классов: 0.98 


In [35]:
from sklearn.metrics import classification_report
print(classification_report(Y_test, LR.predict(X_test)))

              precision    recall  f1-score   support

         ham       0.96      1.00      0.98      1448
        spam       0.98      0.71      0.82       224

    accuracy                           0.96      1672
   macro avg       0.97      0.85      0.90      1672
weighted avg       0.96      0.96      0.96      1672



9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [36]:
mistakes = result[result['Category'] != result['Predicted_Category']]
mistakes

Category Predicted_Category
0                               
2952      ham               spam
881      spam                ham
1961     spam                ham
3864     spam                ham
2575     spam                ham
...       ...                ...
4543     spam                ham
752      spam                ham
309      spam                ham
495       ham               spam
305      spam                ham

[69 rows x 2 columns]

In [39]:
mistakes.merge(spam, left_index=True, right_index=True, left_on=None)

Category_x Predicted_Category Category_y  \
2952        ham               spam        ham   
881        spam                ham       spam   
1961       spam                ham       spam   
3864       spam                ham       spam   
2575       spam                ham       spam   
...         ...                ...        ...   
4543       spam                ham       spam   
752        spam                ham       spam   
309        spam                ham       spam   
495         ham               spam        ham   
305        spam                ham       spam   

                                                Message  
2952                   Hey now am free you can call me.  
881   Reminder: You have not downloaded the content ...  
1961  Guess what! Somebody you know secretly fancies...  
3864  Oh my god! I've found your number again! I'm s...  
2575  Your next amazing xxx PICSFREE1 video will be ...  
...                                                 ...  
4543  FreeMsg Hi baby wow just got a new cam moby. W...  
752   You have an important customer service announc...  
309   TheMob> Check out our newest selection of cont...  
495                    Are you free now?can i call now?  
305   SMS. ac Blind Date 4U!: Rodds1 is 21/m from Ab...  

[69 rows x 4 columns]